server/src/com/cloud/storage/VolumeApiServiceImpl.java

```diff
@@ -2639,22 +2639,23 @@ private int getMaxDataVolumesSupported(UserVmVO vm) {
         return maxDataVolumesSupported.intValue();
     }
 
-    private Long getDeviceId(long vmId, Long deviceId) {
+    private Long getDeviceId(UserVmVO vm, Long deviceId) {
         // allocate deviceId
-        List<VolumeVO> vols = _volsDao.findByInstance(vmId);
+        int maxDataVolumesSupported = getMaxDataVolumesSupported(vm);
```

Comment:
@rafaelweingartner when configuring getMaxDataVolumesSupported(vm) with 6 for the hypervisor of the VM, the VM can have max 6 devices. 1 root (id 0), 1 CD-ROM (id 3) and other 4 for extra disks/volumes.